File: src/java/magmac/api/collect/map/Map.java
package magmac.api.collect.map;

import magmac.api.Option;
import magmac.api.Tuple2;
import magmac.api.iter.Iter;

import java.util.function.Function;
import java.util.function.Supplier;

public interface Map<K, V> {
    V getOrDefault(K key, V other);

    Iter<Tuple2<K, V>> iter();

    Map<K, V> put(K key, V value);

    boolean containsKey(K key);

    V get(K key);

    boolean isEmpty();

    Map<K, V> mapOrPut(K key, Function<V, V> mapper, Supplier<V> supplier);

    Option<Tuple2<Map<K, V>, V>> removeByKey(K key);
}

File: src/web/magmac/api/collect/map/Map.ts
import { Option } from "../../../../magmac/api/Option";
import { Tuple2 } from "../../../../magmac/api/Tuple2";
import { Iter } from "../../../../magmac/api/iter/Iter";
import { Function } from "../../../../java/util/function/Function";
import { Supplier } from "../../../../java/util/function/Supplier";
export interface Map<K,  V> {
	 getOrDefault( key : K,  other : V) : V;
	 iter() : Iter<Tuple2<K, V>>;
	 put( key : K,  value : V) : Map<K, V>;
	 containsKey( key : K) : boolean;
	 get( key : K) : V;
	 isEmpty() : boolean;
	 mapOrPut( key : K,  mapper : Function<V, V>,  supplier : Supplier<V>) : Map<K, V>;
	 removeByKey( key : K) : Option<Tuple2<Map<K, V>, V>>;
}
